### PR TITLE
Add pager to Jobs list

### PR DESCRIPTION
**Pull request by dsam82**

### CODE REVIEW

File: pappl/printer-accessors.c [hunk 1]
```diff
@@ -475,18 +475,29 @@ void
 papplPrinterIterateAllJobs(
     pappl_printer_t *printer,		// I - Printer
     pappl_job_cb_t  cb,			// I - Callback function
-    void            *data)		// I - Callback data
+    void            *data,		// I - Callback data
```
Comment (michaelrsweet) [inline pappl/printer-accessors.c:478]: I'd prefer for this to be first_index (1-based) and limit (number of rows). That matches the IPP terminology.

File: pappl/printer-accessors.c [hunk 2]
```diff
@@ -475,18 +475,29 @@ void
...
 
 
   if (!printer || !cb)
     return;
 
   pthread_rwlock_rdlock(&printer->rwlock);
 
-  for (job = (pappl_job_t *)cupsArrayFirst(printer->all_jobs); job; job = (pappl_job_t *)cupsArrayNext(printer->all_jobs))
-    (cb)(job, data);
+  for (i = 1, job = (pappl_job_t *)cupsArrayFirst(printer->all_jobs); job; ++i, job = (pappl_job_t *)cupsArrayNext(printer->all_jobs))
```
Comment (michaelrsweet) [inline pappl/printer-accessors.c:491]: cupsArrayIndex can be used to start things, then cupsArrayNext to get the subsequent jobs.

File: pappl/printer-accessors.c [hunk 3]
```diff
@@ -475,18 +475,29 @@ void
...
+  {
+    if (nrows == 0)
```
Comment (michaelrsweet) [inline pappl/printer-accessors.c:493]:
```
if (i < limit || limit == 0)
  (cb)(job, data);
else
  break;
```

File: pappl/printer-webif.c [hunk 4]
```diff
@@ -543,14 +545,15 @@ _papplPrinterWebHome(
 
   if (papplPrinterGetNumberOfJobs(printer) > 0)
   {
-    papplClientHTMLPuts(client,
+    papplClientHTMLPrintf(client,
 			"          <table class=\"list\" summary=\"Jobs\">\n"
+      "          <caption>Showing %d most recent jobs</caption>\n"
```
Comment (michaelrsweet) [inline pappl/printer-webif.c:550]: Might as well show the pager links here (which take you to the jobs page) if there are more than N jobs. And make the number of jobs per page a constant that is shared with the jobs page code...

File: pappl/printer-webif.c [hunk 5]
```diff
@@ -572,10 +575,29 @@ void _papplPrinterWebJobs(
...
+      pages = 0;              // total pages
 
   if (!papplClientHTMLAuthorize(client))
   return;
 
+  if (client->operation == HTTP_STATE_GET)
+  {
+    num_form = papplClientGetForm(client, &form);
+
+    if ((value = cupsGetOption("rows", num_form, form)) != NULL)
```
Comment (michaelrsweet) [inline pappl/printer-webif.c:593]: Better to use an index than a page number.

File: pappl/printer.h [hunk 6]
```diff
@@ -469,7 +469,7 @@ extern pappl_system_t	*papplPrinterGetSystem(pappl_printer_t *printer) _PAPPL_PU
 extern pappl_service_type_t papplPrinterGetType(pappl_printer_t *printer) _PAPPL_PUBLIC;
 
 extern void		papplPrinterIterateActiveJobs(pappl_printer_t *printer, pappl_job_cb_t cb, void *data) _PAPPL_PUBLIC;
-extern void		papplPrinterIterateAllJobs(pappl_printer_t *printer, pappl_job_cb_t cb, void *data) _PAPPL_PUBLIC;
+extern void		papplPrinterIterateAllJobs(pappl_printer_t *printer, pappl_job_cb_t cb, void *data, int nrows, int npage) _PAPPL_PUBLIC;
 extern void		papplPrinterIterateCompletedJobs(pappl_printer_t *printer, pappl_job_cb_t cb, void *data) _PAPPL_PUBLIC;
```
Comment (michaelrsweet) [inline pappl/printer.h:473]: Might as well add the first_index and limit to each of the iterators, then you could provide active, completed, and all jobs lists.

**Comment by michaelrsweet** [comment]
@dsam82 I went to look at these changes and they are gone?!?

File: pappl/client-webif.c [hunk 7]
```diff
@@ -371,7 +455,7 @@ papplClientHTMLAuthorize(
 	cupsHashString(auth_hash, sizeof(auth_hash), auth_text, sizeof(auth_text));
 
 	snprintf(cookie, sizeof(cookie), "auth=%s; path=/; expires=%s; httponly; secure;", auth_text, httpGetDateString2(time(NULL) + 3600, expires, sizeof(expires)));
-        httpSetCookie(client->http, cookie);
+        papplClientSetCookie(client, cookie);
```
Comment (michaelrsweet) [inline pappl/client-webif.c:458]: Thinking about this as an outside developer, I won't necessarily know how to properly format the cookie data, so maybe make papplClientSetCookie look like:

```
extern void papplClientSetCookie(pappl_client_t *client, const char *name, const char *value, int expires);
```

where `expires` specifies the number of seconds until expiration or 0 for session cookies (that expire as soon as the browser is closed). The set-cookie code can then do the corresponding snprintf and this code can just be `papplClientSetCookie(client, "auth", auth_text, 3600);`.

File: pappl/printer-accessors.c [hunk 8]
```diff
@@ -450,18 +450,26 @@ void
 papplPrinterIterateActiveJobs(
     pappl_printer_t *printer,		// I - Printer
     pappl_job_cb_t  cb,			// I - Callback function
-    void            *data)		// I - Callback data
+    void            *data,		// I - Callback data
+    int             job_index,    // I - First job index
```
Comment (michaelrsweet) [inline pappl/printer-accessors.c:454]: Make sure to document the job_index starts at 1 in the comment - it'll show up in the generated documentation. Also that a limit of 0 means all jobs...

File: pappl/printer-webif.c [hunk 9]
```diff
@@ -656,7 +663,11 @@ _papplPrinterWebHome(
     pappl_printer_t *printer)		// I - Printer
 {
   ipp_pstate_t		printer_state;	// Printer state
-  char			edit_path[1024];// Edit configuration URL
+  char			edit_path[1024],      // Edit configuration URL
+            buffer[100];          // Limit cookie
+  const int limit = 20;           // Jobs per page
+  int       job_index = 1,        // Job index
+            jobs = 0;             // Total jobs
```
Comment (michaelrsweet) [inline pappl/printer-webif.c:670]: Suggest changing this to "num_jobs" or "njobs" as "jobs" can be confused with printer->jobs...

File: pappl/printer-webif.c [hunk 10]
```diff
@@ -1425,3 +1470,89 @@ time_string(time_t tv,			// I - Time value
 
   return (buffer);
 }
+
```
Comment (michaelrsweet) [inline pappl/printer-webif.c:1473]: Please add descriptive comment explaining the purpose of this function...

File: pappl/style.css [hunk 11]
```diff
@@ -143,6 +143,29 @@ label.image input[type=radio]:checked + img {
   background: rgba(127,127,127,.5);
 }
 
+div.page-container {
```
Comment (michaelrsweet) [inline pappl/style.css:146]: "div.pager" maybe?

Then the link styles can use "div.pager" as the parent class?

**Comment by dsam82** [comment]
@michaelrsweet Are there any other changes that need to be done?

**Comment by michaelrsweet** [comment]
Some minor style things possibly, and I want to look closer at the CSS, but I’ve merged what you have.

**Pull request merged**